### ISSUE TITLE
Scheduled rate limits

### ISSUE THREAD

**Issue by calmh** [syncthing/syncthing#6130]
This has come up a lot in the forum but there's no issue for it here that I can find. We've evolved the rate limits a couple of times and this is something that many/most other rate limiting things have. There is a workaround in using something to periodically reconfigure via the API, but this is not realistically something most users can accomplish.

For reference, here's how this looks in Transmission:

<img width="654" alt="Screen Shot 2019-11-06 at 07 33 22" src="https://user-images.githubusercontent.com/125426/68273970-296f9a00-0068-11ea-9733-2f9d4d1187e2.png">

There's a general limit that can be set or not, and an optional other set of limits that can kick in at certain times. This is both quite flexible and not rocketsciency in UI or implementation. I think we should consider having it.

**Comment by keevBush** [comment]
what is the difference between Speed Limit Mode and Global bandwidth limits????? Is it also necessary to add the limit speed scheduler?

**Comment by calmh** [comment]
In that screenshot, "speed limit mode" is just something you can switch on (or schedule) which switches to a *different* set of rate limits. So if all checkboxes were checked in the screenshot above, we'd have a rate limit of 100/50 normally but then between 04:00 and 14:00 it'd be 10/10 instead.

**Comment by Yithis** [comment]
What do you think should be minimal functionality to pull such feature? Can it be only to toggle on/off speed limit and set there a proper value or should there be also a schedule option?

**Comment by calmh** [comment]
I think there needs to be a schedule for this to be called done

**Comment by Yithis** [comment]
I'm just wondering if it wouldn't be better to divide it into at least two separate Issues/PR. In following days I will try to create a prototype of this feature.

**Comment by calmh** [comment]
I think a partial implementation is fine, it just won't necessarily resolve this issue.

**Comment by Yithis** [comment]
How should scheduler work? I thought about adding separate goroutine in limiter that would try to sleep until change of a limit and then perform needed actions. Does that sound good?

**Comment by calmh** [comment]
Something like that sounds reasonable. Potentially a loop that can sleep waiting for either timer events or manual commands on a channel (for "change to speed limited mode now" kind of things).

**Comment by uok** [comment]
Sorry for a noob question: can this timer be made universal, e.g. also be used for another timer? There was this idea floating around (idk if there is an issue, if not I will create it) that folders/devices can be paused for a certain time.

**Comment by calmh** [comment]
@uok wrong ticket. (Timers are cheap, we can have as many as we like.)

**Comment by calmh** [comment]
The cron package might be useful too. https://godoc.org/gopkg.in/robfig/cron.v2

**Comment by kotoko** [comment]
May I suggest a little different approach? Add API endpoints to syncthing that allows others programs change rate limits "on the fly". This should not make syncthing to drop current connections but gracefully throttle up/down bandwidth. This allows users to create bash/python scripts to be run from cron. Someone else could also create separate GUI program that would allow advanced scheduling based on time & date and maybe others factors too.

I didnt't think much about technical details but I have 2 ideas.
1. Create web endpoint and protect it with current user + password combo. It's problematic when you want automatize because you need to save password with plain text inside script.
2. Allow only programs that are run on the same machine by the same user. Maybe FIFO file or something like that?

It's just a suggestion.

**Comment by calmh** [comment]
Rate limits can already be changed on the fly using the config API, without affecting existing connections.

**Comment by gitbxo** [comment]
@calmh Using the cron package sounds like a great idea.
Can I consider working on this ticket?
Can partial implementations be implemented and merged if they don't break anything?
Example: partial implementation means if the backend is implemented without the GUI changes and the GUI changes are made as a separate pull request?

**Comment by calmh** [comment]
@gitbxo Sure. I think backend and GUI can be sensibly separated into separate PRs, otherwise we don't generally merge partial stuff, no.

**Comment by AudriusButkevicius** [comment]
Given we now have the CLI packed with syncthing. and based on what I am hearing here, are planning to use a cron notation for these schedules (something like `0 8 * * * set ratelimit 400kb` etc, isn't it now equivalent to just using actual cron+cli?

**Comment by gitbxo** [comment]
@calmh @AudriusButkevicius 

Looking into the code, it seems that the code may be fetching the rate limit multiple times.
Assuming that is true, we don't really need to use cron at all.

lib/connections/limiter.go function getRateLimiter already fetches the rateLimiter.
Looks like that can be updated to check the cron schedule before returning the rateLimiter.

If that makes sense, then we would not need to actually use cron or anything, just return the correct rate limit based on the current time.

What do you think?

**Comment by calmh** [comment]
I don't think the existence of the cli changes the equation much here, this is a still a commonly requested feature that fits internally and a lot of users don't have cron+cli or an easy equivalent, or won't know what to do with it if they do.

getRateLimiter is called to get a limited when a connection is established, iirc. A connection is long lived. So limiters will need to be adjusted on the fly. I think `(*limiter)CommitConfiguration()` might be a good thing to look at for inspiration.

**Comment by AudriusButkevicius** [comment]
Right, but is this being implemented as a cron notation? Because if yes, they won't know what to do with a cron editor inside of syncthing either.

**Comment by calmh** [comment]
Certainly no. I think that's just mentioned as some sort of convenient clock-time timer firing library. I'd suggest something closer to the original screenshot I posted above.

**Comment by gitbxo** [comment]
@calmh @AudriusButkevicius 

I agree that we don't want to have user's editing some cron string monstrosity.
That is why separating the backend functionality from the UI changes will help ensure:
(a) The backend is following the right pattern and methodology
(b) User / GUI discussions and backend technology decisions can be handled separately.

Phase I - backend:
Use cron-string format to support selecting days of month or days of week in a well-known format.
Need not support entire cron-format but the format is well-known, simple, and flexible enough for what we need.
That makes the format easy to understand for developers.

Phase II - GUI:
This will design the GUI for what is exposed to the users and will create the required cron strings.
Agreed that we don't want user's to be editing the cron strings, so the code will create the cron string from the GUI options.
The GUI should basically allow the users to input one or more of the following:
* days of month, start/stop times and rate limits
* days of week, start/stop times and rate limits

It appears that getRateLimiter is called only when initializing the connection, so if there is a scheduled time, then we can simply add a timer to update the returned rate limiter whenever the scheduled time is crossed.
If you are fine with that, then I can start with the backend work for Phase I.

**Comment by pleriche** [comment]
This would be an incredibly useful feature for users using syncthing between work sites that are very busy during office hours, and where synchronization should thus ideally only happen after hours. Not everyone is tech savvy enough to be able to use the REST api.

Are there still plans to implement this?

**Comment by tomasz1986** [comment]
The feature request is still open, so if someone skilled enough writes the code and prepares a proper pull request, I'm sure it will be considered and reviewed.

**Comment by jsmdy** [comment]
same here, one year later
